### PR TITLE
Issue #137: Java Test Sample - Async Lambda DynamoDB

### PR DESCRIPTION
**Pull request by gmartim**
Issue #137

Implementing the test described on the issue and detailed on Quip.

This test intent to check an async architecture using S3, lambda and DynamoDB.

### CODE REVIEW

File: java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        PutObjectResponse putObjectResponse = s3Client.putObject(putObjectRequest, RequestBody.fromString(MESSAGE));
+        assertNotNull(putObjectResponse, "putObjectResponse may not be null");
+
+        if (recordTransformationTable != null) {
+            try {
+                Callable<String> callable = getContentFromDynamoDB(recordTransformationTable, key);
+
+                String contentFromRecordTransformationTable = callable.call();
+
+                assertEquals(contentFromRecordTransformationTable, contentFromRecordTransformationTable,
```
Comment (poprahul) [inline java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java:171]: Not sure if I follow this line. Expected and Actual values are the same here in assertEquals?

File: java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                assertTrue((exception instanceof RecordNotFoundException), "An exception occured and shoud not have");
+            }
+        }
+    }
+
+    private Callable<String> getContentFromDynamoDB(String tableName, String key)
+            throws InternalErrorException, RecordNotFoundException {
+        IntervalFunction intervalFunction = ofExponentialBackoff(3000l, 1.5d);
+
+        RetryConfig retryConfig = RetryConfig.custom().maxAttempts(3).intervalFunction(intervalFunction)
```
Comment (poprahul) [inline java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java:207]: Can we not override dynamodb client config instead to handle retry instead resiliency4j library? like below -

/**
     * Method to overwrite the default SDK client configuration behavior
     *
     * @return ClientConfiguration with custom timeout values and retry method
     */
    private static ClientConfiguration createDynamoDBClientConfiguration() {
        ClientConfiguration clientConfiguration = new ClientConfiguration()
               .withConnectionTimeout(MyDynamoDBClientParameters.connectionTimeout)
                .withClientExecutionTimeout(MyDynamoDBClientParameters.clientExecutionTimeout)
                .withRequestTimeout(MyDynamoDBClientParameters.requestTimeout)
                .withSocketTimeout(MyDynamoDBClientParameters.socketTimeout)
                .withRetryPolicy(PredefinedRetryPolicies
                        .getDynamoDBDefaultRetryPolicyWithCustomMaxRetries(
                                        MyDynamoDBClientParameters.maxErrorRetries));

        return clientConfiguration;
    }
}

**Review by poprahul — COMMENTED** [review]
Pretty clean code. Just few minor comments on the test class.

File: java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java (hunk 2 above)
Comment (dmahapatro) [reply java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java:207]: @poprahul Isn't DDB client config retry for the client's connection? This test is meant to retry until it successfully retrieves an item from the table.

Comment (gmartim) [reply java-test-samples/async-lambda-dynamodb/src/test/java/com/amazon/aws/sample/TestAsyncTransformation.java:207]: Yes, it is like a wait until get the flow complete and the record inserted on the table.

**Comment by poprahul** [comment]
Thanks for fixing the test case. Looks good.

**Review by dmahapatro — APPROVED** [review]

**Pull request merged**